src/cluster_slot_stats.c

```diff
@@ -0,0 +1,173 @@
+    slotStatEntry entry_a = *((slotStatEntry *) a);
+    slotStatEntry entry_b = *((slotStatEntry *) b);
+    return entry_b.stat - entry_a.stat;
+}
+
+static void collectAndSortSlotStats(slotStatEntry slot_stats[], int order_by, int desc) {
+    int i = 0;
+    
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToMyShard(slot)) {
+            slot_stats[i].slot = slot;
+            slot_stats[i].stat = getSlotStat(slot, order_by);
+            i++;
+        }
+    }
+    qsort(slot_stats, i, sizeof(slotStatEntry), (desc) ? slotStatEntryDescCmp : slotStatEntryAscCmp);
+}
+
+static void addReplySlotStat(client *c, int slot) {
+    addReplyLongLong(c, slot);
+    addReplyMapLen(c, 1);
+    addReplyBulkCString(c, "key-count");
+    addReplyLongLong(c, countKeysInSlot(slot));
+}
+
+static void addReplySlotStats(client *c, unsigned char *assigned_slots, int startslot, int endslot, int len) {
+    addReplyMapLen(c, len);
+
+    for (int slot = startslot; slot <= endslot; slot++) {
+        if (assigned_slots[slot]) addReplySlotStat(c, slot);
+    }
+}
+
+static void addReplySortedSlotStats(client *c, slotStatEntry slot_stats[], long limit) {
+    int num_slots_assigned = getMyShardSlotCount();
+    int len = min(limit, num_slots_assigned);
+    addReplyMapLen(c, len);
+
+    for (int i = 0; i < len; i++) {
+        addReplySlotStat(c, slot_stats[i].slot);
+    }
+}
+
+static void sortAndAddReplySlotStats(client *c, int order_by, long limit, int desc) {
+    slotStatEntry slot_stats[CLUSTER_SLOTS];
+    collectAndSortSlotStats(slot_stats, order_by, desc);
+    addReplySortedSlotStats(c, slot_stats, limit);
+}
+
+void clusterSlotStatsCommand(client *c) {
+    if (server.cluster_enabled == 0) {
+        addReplyError(c,"This instance has cluster support disabled");
+        return;
+    }
+
+    /* Parse additional arguments. */
+    if (c->argc == 5 && !strcasecmp(c->argv[2]->ptr,"slotsrange")) {
+        /* CLUSTER SLOT-STATS SLOTSRANGE start-slot end-slot */
+        int startslot, endslot;
+        if ((startslot = getSlotOrReply(c,c->argv[3])) == C_ERR ||
+            (endslot = getSlotOrReply(c,c->argv[4])) == C_ERR) {
+            return;
+        }
+        if (startslot > endslot) {
+            addReplyErrorFormat(c,"Start slot number %d is greater than end slot number %d", startslot, endslot);
+            return;
+        }
+        /* Initialize slot assignment array. */
+        unsigned char assigned_slots[CLUSTER_SLOTS]= {UNASSIGNED_SLOT};
+        int len = 0;
+        markSlotsAssignedToMyShard(assigned_slots, startslot, endslot, &len);
+        addReplySlotStats(c, assigned_slots, startslot, endslot, len);
+
+    } else if (c->argc >= 4 && !strcasecmp(c->argv[2]->ptr,"orderby")) {
+        /* CLUSTER SLOT-STATS ORDERBY metric [LIMIT limit] [ASC | DESC] */
+        int desc = 1, order_by = INVALID;
+        if (!strcasecmp(c->argv[3]->ptr, "key-count")) {
```

Comment:
Here, a little bit confused me.  I know we want to get metric of key-count. cpu-usec, and memory-bytes through the command CLUSTER SLOT-STATS , but the comment "CLUSTER SLOT-STATS ORDERBY metric [LIMIT limit] [ASC | DESC]" make reviewer confused. I am not sure if you want to implement these 3 parts metric in this pr or you have plan to implement the  cpu-usec, and memory-bytes in 2 separated prs

If you want to implement them in one pr, how about the command format as CLUSTER SLOT-STATS ORDERBY [key-count] [cpu-used] [memory-bytes] [LIMIT limit] [ASC | DESC]

If you just want to implement key-count here, i suggest the command format as CLUSTER SLOT-STATS ORDERBY key-count [LIMIT limit] [ASC | DESC]

Reply:
Thanks for clearing this. Let us simply keeping this for saving your energy.